internal/processor/expfilter/matcher.go

```diff
@@ -0,0 +1,84 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expfilter
+
+import (
+	"log"
+
+	"github.com/antonmedv/expr"
+	"github.com/antonmedv/expr/vm"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+type Matcher struct {
+	program *vm.Program
+	v       vm.VM
+}
+
+type env struct {
+	MetricName string
+	LabelKey   string
+	LabelVal   string
+}
+
+func NewMatcher(searchStr string) (*Matcher, error) {
+	program, err := expr.Compile(searchStr)
+	if err != nil {
+		return nil, err
+	}
+	return &Matcher{program: program, v: vm.VM{}}, nil
+}
+
+func (m *Matcher) MatchMetric(metric pdata.Metric) bool {
+	metricName := metric.Name()
+	tested := false
+	switch metric.DataType() {
+	case pdata.MetricDataTypeIntGauge:
+		pts := metric.IntGauge().DataPoints()
+		for i := 0; i < pts.Len(); i++ {
+			pt := pts.At(i)
+			found := false
+			pt.LabelsMap().ForEach(func(k string, v pdata.StringValue) {
+				tested = true
+				if m.match(env{MetricName: metricName, LabelKey: k, LabelVal: v.Value()}) {
```

Comment:
Probably for performance there needs to be no more than one match call per Datapoint, not one per label.  Also having one per datapoint means you can do things like consider multiple labels in one expression.

Reply:
Agree, but wasn't sure how to go about that. Thinking about it some more, maybe it's possible to just make the label map available to expr. Will try that out.

Reply:
You can define helper functions that would access labels in a simple manner, without having to convert it to a map.